src/services/completions.ts

```diff
@@ -1667,134 +1687,52 @@ namespace ts.Completions {
-                const isDefaultExport = symbol.escapedName === InternalSymbolName.Default;
-                const nonLocalSymbol = symbol;
-                if (isDefaultExport) {
-                    symbol = getLocalSymbolForExportDefault(symbol) || symbol;
-                }
-                if (typeChecker.isUndefinedSymbol(symbol)) {
-                    return;
-                }
-                const original = skipAlias(nonLocalSymbol, typeChecker);
-                const symbolName = getNameForExportedSymbol(symbol, target);
-                const existingSuggestions = results.get(getSymbolId(original));
-                if (!some(existingSuggestions, s => s.symbolName === symbolName && moduleSymbolsAreDuplicateOrigins(moduleSymbol, s.origin.moduleSymbol))) {
-                    const origin: SymbolOriginInfoExport = {
-                        kind: SymbolOriginInfoKind.Export,
-                        moduleSymbol,
-                        isDefaultExport,
-                        isFromPackageJson,
-                        exportName,
-                        fileName: file?.fileName
-                    };
-                    results.add(getSymbolId(original), {
-                        symbol,
+        /** Mutates `symbols`, `symbolToOriginInfoMap`, and `symbolToSortTextMap` */
+        function collectAutoImports(resolveModuleSpecifiers: boolean) {
+            if (!shouldOfferImportCompletions()) return;
+            Debug.assert(!detailsEntryId?.data);
+            const start = timestamp();
+            host.log?.(`collectAutoImports: starting, ${resolveModuleSpecifiers ? "" : "not "}resolving module specifiers`);
+            const lowerCaseTokenText = previousToken && isIdentifier(previousToken) ? previousToken.text.toLowerCase() : "";
+            const exportInfo = codefix.getSymbolToExportInfoMap(sourceFile, host, program, /*filterByPackageJson*/ !detailsEntryId);
+            exportInfo.forEach((info, key) => {
+                const [symbolName] = key.split("|");
+                if (!detailsEntryId && isStringANonContextualKeyword(symbolName)) return;
+                const isCompletionDetailsMatch = detailsEntryId && some(info, i => detailsEntryId.source === stripQuotes(i.moduleSymbol.name));
+                if (isCompletionDetailsMatch || stringContainsCharactersInOrder(symbolName.toLowerCase(), lowerCaseTokenText)) {
+                    // If we don't need to resolve module specifiers, it doesn't matter which SymbolExportInfo
+                    // we use. Each is importable by the same name and resolves to the same declaration.
+                    const { moduleSpecifier, exportInfo } = resolveModuleSpecifiers
+                        ? codefix.getBestImportFixForExports(info, sourceFile, /*position*/ undefined, /*preferTypeOnlyImport*/ false, /*useRequire*/ false, program, host, preferences)
+                        : { moduleSpecifier: undefined, exportInfo: info[0] };

```

Comment:
This is the real crux of the PR. In `getSymbolsFromOtherSourceFileExports`, we deduplicated re-exports on the fly, since they don’t matter until we need module specifiers. But for import statement completions, we need to resolve module specifiers up front (for reasons discussed beginning around https://github.com/microsoft/TypeScript/issues/31658#issuecomment-738455502), so `getSymbolsFromOtherSourceFileExports` was utterly unusable.

Now, instead of tossing out re-exports, we let `codefix.getSymbolToExportInfoMap` group exports (the grouping strategy is discussed in an earlier review comment). Then, if we are being asked to eagerly resolve module specifiers, we can send all the re-exports in the group to `codefix.getBestImportFixForExports`, and if we are just being asked to show an auto-import completion as we’ve always done, we simply pick the first export in the group, since they’re interchangeable until we need module specifiers. This lets us use 90% of the same code between import statement completions (the new feature in this PR) and regular auto imports.